### DISCUSSION TITLE
How to use multiple MySQL databases with prisma?

### DISCUSSION THREAD

**Discussion by adornerz** [prisma/prisma#11324] [Q&A] [answered]
I'm trying to implement Prisma to my existing `Next.js` with `MySQL` project, but got stuck.
The way my project works is basicially I access different databases based on the url. For example: `site.com/proj1` needs to access `proj1` database, and `site.com/proj2` needs to access `proj2`, `/admin` needs to access the admin database and so on. All databases have the exact same set of tables and can be accessed on the same port.
The way I do this is using `mysql2`, and create a connection like this:
```js
const connection = mysql.createConnection({
    'host'      : 'localhost',
    'port'      : '3306',
    'user'      : 'root',
    'password'  : 'samplepass'
});
```
Then use the database needed with:
```js
useDB: async function (DBName, con = connection) {
       await con.promise().query(`USE ${DBName}`);
}
```
Can this be done with Prisma and if it's not a feature, is there any workaround to do this?

**Comment by austincrim** [comment, marked as answer]
Hey @adornerz 👋 ,

You can programmatically override the connection URL using when you instantiate the Prisma Client using the [`datasource` option](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#datasources).

```ts
import { PrismaClient } from '@prisma/client'

const client1 = new PrismaClient({ datasources: { db: { url: 'some connection string' } } })
const client2 = new PrismaClient({ datasources: { db: { url: 'different connection string' } } })
```

We also have an existing issue (#2443) that contains a lot of other info and prior Prisma thought on this topic, you might be interested to skim through that as well.

Let me know if I can clarify anything further!

**Reply by adornerz** [reply]
Hey @austincrim , thanks for taking your time to help.
So, if I understood this correctly, I can do something like this:
For `site.com/dbName`, where `dbName` is saved in a variable `dbToAccess`:
```js
const client1 = new PrismaClient({ datasource: { db: { url: 'mysql://user:pass@localhost:port/' + dbToAccess } } })
```
This will connect me to the desired database, and then I just continue from there?

EDIT: Tried that, but I got `SyntaxError: Cannot use import statement outside a module` because the file I'm working on is a script file, not a module, so I used `const { PrismaClient } = require('@prisma/client')`, but here I got the `Unknown property datasource provided to PrismaClient constructor. Did you mean "datasources"?` error from prisma. What can I do?

**Reply by austincrim** [reply]
@adornerz I apologize, the option is actually called `datasources` with an **s**.

**Reply by adornerz** [reply]
@austincrim that seems to work!
Thank you very much for helping!

**Comment by ghost** [comment]
@austincrim Hey, The above solution worked but When I use `prisma db pull` and `prisma generate` the schema file is updated based on the default DB.

```javascript 
/* schema.prisma */
datasource db {
  provider = "mysql"
  url      = env("DATABASE_URL")
}
```
```javascript
/* .env */
DATABASE_URL=mysql://user:pass@host/db
```
```javascript
/* Initializing */
const client1 = new PrismaClient({ datasources: { db: { url: 'mysql://user:pass@host/db' } } })
const client2 = new PrismaClient({ datasources: { db: { url: 'mysql://user:pass@diff-host/diff-db' } } })
```

**Reply by max-d3v** [reply]
Hey! found a solution? im having the same problem with the generated client, it wont override properly the connection even when passing url like you did

**Reply by pvshum** [reply]
Hey @max-d3v have you tried using `--schema` flag?

E.g. following works for me.

I've created a separate schema file in separate dir.
Then to generate types, apply migrations, launch studio, etc I just use  `--schema` flag

```bash
npx prisma generate --schema prismaReporting/schema.prisma
```
Then in the code I use it as follows:

```javascript
import { PrismaClient as PrismaReportingClient, type ReportEntry } from '@prismaReporting/client';

const prismaReporting = new PrismaReportingClient();
...
await prismaReporting.reportEntry.create({
   data,
});
```

Let me know if it works for you?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
